### PR TITLE
V1 emails

### PR DESCRIPTION
**Pull request by mbuchel**
ready for merging with the rest

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -102,6 +107,34 @@
...
+            <artifactId>regions</artifactId>
+            <version>2.11.2</version>
+        </dependency>
+        <dependency>
+            <groupId>javax.mail</groupId>
+            <artifactId>mail</artifactId>
+            <version>1.4</version>
+            <exclusions>
+                <exclusion>
+                    <groupId>jakarta.activation</groupId>
```
Comment (toadzky) [inline pom.xml]: javax.mail doesn't use jakarta, it uses the old one. exclude the old one here

File: pom.xml [hunk 2]
```diff
@@ -45,6 +46,10 @@
                         <groupId>com.sun.activation</groupId>
                         <artifactId>jakarta.activation</artifactId>
                     </exclusion>
```
Comment (toadzky) [inline pom.xml]: stop excluding this one

File: pom.xml [hunk 3]
```diff
@@ -102,6 +107,34 @@
             <version>8.10.2</version>
         </dependency>
 
+        <!-- AWS Sending Emails -->
+        <dependency>
+            <groupId>software.amazon.awssdk</groupId>
+            <artifactId>ses</artifactId>
+            <version>2.11.2</version>
```
Comment (toadzky) [inline pom.xml]: use the aws sdk bom to manage these versions

File: src/main/kotlin/io/zerobase/smarttracing/GraphDao.kt [hunk 4]
```diff
@@ -150,7 +150,7 @@ class GraphDao(private val driver: Driver, val phoneUtil: PhoneNumberUtil) {
                 txn.run(
                         """
                         MATCH (o:Organization { id: '${id.value}' })
-                        SET o.multisite = 'true'
+                        SET o.multisite = '${state}'
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/GraphDao.kt]: just go ahead and revert the unrelated dao changes. all of this is changing with the neptune move anyway

File: src/main/kotlin/io/zerobase/smarttracing/GraphDao.kt [hunk 5]
```diff
@@ -297,17 +318,19 @@ class GraphDao(private val driver: Driver, val phoneUtil: PhoneNumberUtil) {
      *
      * @param id id of the user to delete
      */
-    fun deleteUser(id: UserId) {
-        driver.session().use {
+    fun deleteUser(id: UserId): String {
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/GraphDao.kt]: don't make the delete return a piece of the user. either return the whole user or just do a get first

File: src/main/kotlin/io/zerobase/smarttracing/Main.kt [hunk 6]
```diff
@@ -19,7 +19,8 @@ import javax.servlet.FilterRegistration
 
 typealias MultiMap<K,V> = Map<K, List<V>>
 
-data class Config(val database: Neo4jConfig, val siteTypeCategories: MultiMap<String, String>, val scannableTypes: List<String>): Configuration()
+data class AmazonConfig(val from: String)
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/Main.kt]: what does "from" mean? it especially doesn't mean anything in the context of Amazon. aws config isn't the same as email config. split email from aws and give the fields useful names

File: src/main/kotlin/io/zerobase/smarttracing/Main.kt [hunk 7]
```diff
@@ -19,7 +19,8 @@ import javax.servlet.FilterRegistration
...
+data class Config(val database: Neo4jConfig, val amazonaws: AmazonConfig, val siteTypeCategories: MultiMap<String, String>, val scannableTypes: List<String>): Configuration()
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/Main.kt]: just name the field `aws`

**Review by toadzky — CHANGES_REQUESTED** [review]

File: src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+import java.io.IOException
+import java.nio.ByteBuffer
+import java.util.Properties
+import software.amazon.awssdk.core.SdkBytes
+import software.amazon.awssdk.services.ses.model.SendRawEmailRequest
+import software.amazon.awssdk.services.ses.model.RawMessage
+import io.zerobase.smarttracing.AmazonConfig
+
+// Currently only implements the AmazonSES
+class Amazon(private val config: AmazonConfig) {
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt]: no we don't want a class taht represents all of amazon. we want an abstraction over sending emails, that's all. if we start using other services in aws, we will write one for that service at that time

File: src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+    private val session: Session
+    private val client: SesClient
+
+    init {
+        session = Session.getDefaultInstance(Properties())
+        client = SesClient.builder().region(Region.US_EAST_1).build()
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt]: still no. this needs to be passed in. read up on the dependency inversion principle

File: src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt (hunk 8 above)
Comment (toadzky) [reply src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt]: also read up on the law of demeter. don't pass in the config when you need the fields. pass in the fields you need

File: src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+     * @param subject subject to send this under
+     * @param html html to send
+     * @param to address to send to
+     * @param url the url to encode in a qr code and send
+     *
+     * @throw AddressException when the address is incorrect
+     * @throw MessagingException when we cannot message
+     * @throw IOException input output exception
+     */
+    private fun sendQRScannable(subject: String, html_string: String, to: String, url: String) {
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt]: the email abstraction shouldn't know anything about qr codes

File: src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+    /**
+     * Sends thank you for joining email to an organization
+     *
+     * @param to address to send to
+     *
+     * @throw AddressException when the address is incorrect
+     * @throw MessagingException when we cannot message
+     * @throw IOException input output exception
+     */
+    fun scannableOrgEmail(to: String, url: String) {
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt]: none of these function names describe the action being taken. on top of that, the contents shouldn't be in the email abstraction. at the very least, they should be templates stored as resources.

**Review by toadzky — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 2 above)
Comment (mbuchel) [reply pom.xml]: failure building if i remove this

File: src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt (hunk 8 above)
Comment (mbuchel) [reply src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt]: i thought yesterday we agreed that we wanted it to be an abstraction of the entire AWS services? but either way this is resolved now

File: src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt (hunk 10 above)
Comment (mbuchel) [reply src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt]: it needs to be able to handle qr codes as an attachment, as such i thought it would make more sense for this section to generate the pdf here as opposed to passing it from function to function

File: src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt (hunk 11 above)
Comment (mbuchel) [reply src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt]: what names do you suggest? i am bad at naming, i am moving the templates into resources now

File: src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt (hunk 10 above)
Comment (toadzky) [reply src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt]: no it needs to be able to send attachments. QR codes are an attachment. the email abstraction should *not* know about what the attachments are. it should take at most an address, a template reference to render, the model to render, and any attachments.

Comment (mbuchel) [reply src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt]: alright i will set that up

File: src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt (hunk 11 above)
Comment (toadzky) [reply src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt]: for now, let's keep the email abstraction simple: it takes an address, a body, and an optional attachment. we can make another `NotificationManager` class to abstract whether a user is notified by email or sms or push notification.

Comment (mbuchel) [reply src/main/kotlin/io/zerobase/smarttracing/resources/Amazon.kt]: alright will push that soon

File: pom.xml [hunk 12]
```diff
@@ -107,29 +110,36 @@
...
 
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.sun.activation</groupId>
+            <artifactId>jakarta.activation</artifactId>
+            <version>1.2.1</version>
+        </dependency>
```
Comment (toadzky) [inline pom.xml]: do we actually need this? i don't think we use it explicity

File: pom.xml [hunk 13]
```diff
@@ -107,29 +110,36 @@
             <version>8.10.2</version>
         </dependency>
 
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+        </dependency>
```
Comment (toadzky) [inline pom.xml]: nit: this should be under `Other Dependencies`

File: src/main/kotlin/io/zerobase/smarttracing/GraphDao.kt [hunk 14]
```diff
@@ -150,7 +150,7 @@ class GraphDao(private val driver: Driver, val phoneUtil: PhoneNumberUtil) {
                 txn.run(
                         """
                         MATCH (o:Organization { id: '${id.value}' })
-                        SET o.multisite = '${state}'
+                        SET o.multisite = 'true'
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/GraphDao.kt]: no, this was correct. revert your change.

File: src/main/kotlin/io/zerobase/smarttracing/Main.kt [hunk 15]
```diff
@@ -18,9 +19,10 @@ import javax.servlet.DispatcherType
 import javax.servlet.FilterRegistration
 
 typealias MultiMap<K,V> = Map<K, List<V>>
+typealias EmailConfig = Map<EmailType, EmailConfigEntry>
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/Main.kt]: you can import this type alias from the models file instead of redefining it

File: src/main/kotlin/io/zerobase/smarttracing/Main.kt [hunk 16]
```diff
@@ -18,9 +19,10 @@ import javax.servlet.DispatcherType
...
 
-data class AmazonConfig(val from: String)
-data class Config(val database: Neo4jConfig, val amazonaws: AmazonConfig, val siteTypeCategories: MultiMap<String, String>, val scannableTypes: List<String>): Configuration()
+data class AWS(val fromEmailSES: String)
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/Main.kt]: i think you misunderstood a little bit. the email we send from isn't SES or AWS config. it's just app config. SES config would be things like which region to use or an alternate endpoint instead of AWS to use (useful for testing). i would structure it something like this:

```kotlin
data class SimpleEmailService(val region: String = 'us-east-1', val endpoint: URI? = null)
data class AwsConfig(val ses: SimpleEmailService = SimpleEmailService())
```

that way, we can add other service configs into the aws config when we need to. the yaml would look like this:

```yaml
aws:
  ses:
    region: us-east-1
```

File: src/main/kotlin/io/zerobase/smarttracing/models/Exceptions.kt [hunk 17]
```diff
@@ -1,3 +1,4 @@
 package io.zerobase.smarttracing.models
 
 class InvalidPhoneNumberException(message: String) : Exception(message)
+class InvalidEmailType(message: String) : Exception(message)
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/models/Exceptions.kt]: style nit: suffix exception class names with `Exception`

File: src/main/kotlin/io/zerobase/smarttracing/models/Models.kt [hunk 18]
```diff
@@ -20,3 +20,16 @@ data class Location(
...
 )
+
+data class EmailConfigEntry(
+    val subject: String,
+    val resource: String,
+    val attach_name: String?
+)
+
+enum class EmailType {
+    CREATE_ORG, CREATE_SCANNABLE,
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/models/Models.kt]: nit: just spell out `ORGANIZATION`, no reason to shorten it

File: src/main/kotlin/io/zerobase/smarttracing/resources/AmazonSES.kt [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+import software.amazon.awssdk.core.SdkBytes
+import software.amazon.awssdk.services.ses.model.SendRawEmailRequest
+import software.amazon.awssdk.services.ses.model.RawMessage
+
+class AmazonSES : EmailSender {
+    private val client: SesClient
+
+    init {
+        client = SesClient.builder().region(Region.US_EAST_1).build()
+    }
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/resources/AmazonSES.kt]: you are still missing the point on the dependency inversion principle. classes should *not* create what they need, they should be given it. do it like this:

```kotlin
class AmazonEmailSender(private val client: SesClient) {
...
}

File: src/main/kotlin/io/zerobase/smarttracing/resources/Email.kt [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+import java.nio.ByteBuffer
+import java.util.Properties
+import java.io.File
+import java.io.InputStream
+import com.google.common.io.Resources
+import javax.activation.DataHandler
+
+// Currently only implements the AmazonSES
+class Email(private val fromEmailSES: String, private val amazonSES: AmazonSES,
+            private val config: EmailConfig) {
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/resources/Email.kt]: a) don't include `SES` in variable names for other classes
b) the name of this class suggests that it represents an actual email object, but it seems like it's a wrapper for sending emails
c) the point of an abstraction is to use the abstraction. why are you passing in `AmazonSES` instead of `EmailSender`

File: src/main/kotlin/io/zerobase/smarttracing/resources/EmailSender.kt [hunk 21]
```diff
@@ -0,0 +1,5 @@
+package io.zerobase.smarttracing.resources
+
+interface EmailSender {
+    fun sendEmail(arr: ByteArray)
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/resources/EmailSender.kt]: this interface is too coupled to exactly how SES works. the point of an abstraction is to abstract things. this should probably be

```kotlin
fun sendEmail(toAddress: String body: EmailTemplate, attachment = Array<byte>? = null)
```

where `EmailTemplate` is something like

```kotlin
sealed class EmailTemplate {
    abstract fun byte[] render()
}
...
```

we could also do something like 
```kotlin
fun sendEmail(toAddress: String, body: String, contentType: String = "text/html", attachment: Array<Byte>? = null)
```

File: src/main/kotlin/io/zerobase/smarttracing/resources/OrganizationsResource.kt [hunk 22]
```diff
@@ -51,28 +52,31 @@ data class SiteResponse(val id: String, val name: String)
...
+                            private val email: Email) {
 
     @POST
     @Creator
     @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE", justification = "false positive")
     fun createOrganization(request: CreateOrganizationRequest): IdWrapper? {
         val name = request.name
         val phone = request.contact.phone
-        val email = request.contact.email
+        val to_email = request.contact.email
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/resources/OrganizationsResource.kt]: a) style: `camelCase` not `snake_case`
b) naming: `organizationEmail` not `toEmail`

File: src/main/kotlin/io/zerobase/smarttracing/resources/OrganizationsResource.kt [hunk 23]
```diff
@@ -138,7 +142,8 @@ class OrganizationsResource(val dao: GraphDao, private val siteTypes: MultiMap<S
 
         val id = dao.createScannable(OrganizationId(orgId), SiteId(siteId), type, singleUse)
 
-        id?.let { amazon.scannableOrgEmail(dao.getEmailOrg(OrganizationId(orgId)), it.value) }
+        // TODO: Change null into bytearray for pdf
+        id?.let { email.send(dao.getEmailOrg(OrganizationId(orgId)), null, EmailType.CREATE_SCANNABLE) }
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/resources/OrganizationsResource.kt]: this dao method should me renamed. at the least, it should be `getOrgEmail`, but there's not a great reason to shorten it so `getOrganizationEmail` would be better

File: src/main/kotlin/io/zerobase/smarttracing/resources/UsersResource.kt [hunk 24]
```diff
@@ -39,8 +40,8 @@ class UsersResource(val dao: GraphDao, val amazon: Amazon) {
                 request.name, request.contact.phone,
                 request.contact.email, DeviceId(request.deviceId)
             )
-            val email = request.contact?.email
-            email?.let { amazon.thanksForJoiningUser(it) }
+            val to_email = request.contact?.email
```
Comment (toadzky) [inline src/main/kotlin/io/zerobase/smarttracing/resources/UsersResource.kt]: same comment as above

File: src/main/resources/config.yml [hunk 25]
```diff
@@ -9,7 +9,22 @@ database:
...
 amazonaws:
-  from: ${FROM_EMAIL:-""}
+  fromEmailSES: ${FROM_EMAIL:-""}
+
+emailConfig:
+  CREATE_ORG:
+    subject: "Thank you joining our program"
+    resource: "email-templates/org-join"
+  CREATE_SCANNABLE:
+    subject: "Thank you for creating a scannable"
```
Comment (toadzky) [inline src/main/resources/config.yml]: scannable is an internal term, it's not part of what we are communicating to users. we should have official copy eventually, but for now, just replace scannable with "QR Code". we'll update it later

**Review by toadzky — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 12 above)
Comment (mbuchel) [reply pom.xml]: we use it explicitly, in getting the data handler class

File: src/main/kotlin/io/zerobase/smarttracing/GraphDao.kt (hunk 14 above)
Comment (mbuchel) [reply src/main/kotlin/io/zerobase/smarttracing/GraphDao.kt]: See your previous comment regarding reverting back the fix as we are switching databases

File: src/main/kotlin/io/zerobase/smarttracing/Main.kt (hunk 15 above)
Comment (mbuchel) [reply src/main/kotlin/io/zerobase/smarttracing/Main.kt]: my bad this was supposed to be deleted from the pushed code

File: src/main/kotlin/io/zerobase/smarttracing/resources/Email.kt (hunk 20 above)
Comment (mbuchel) [reply src/main/kotlin/io/zerobase/smarttracing/resources/Email.kt]: what name do you suggest for this email class? this class creates the email object and passes the raw byte array into the object that is sending the email

File: src/main/kotlin/io/zerobase/smarttracing/resources/EmailSender.kt (hunk 21 above)
Comment (mbuchel) [reply src/main/kotlin/io/zerobase/smarttracing/resources/EmailSender.kt]: we will have to talk about this because i am confused here, this interface is only in case we want to change the underlying technology used to send the email after it is already packed into the binary state, the email class packs the data into the binary state and sends it to the email sender

Comment (toadzky) [reply src/main/kotlin/io/zerobase/smarttracing/resources/EmailSender.kt]: no, the interface is *not* to change tech for sending after it's already packed into a binary state. the whole point is to be able to change out email providers. if we switch to something like mandril, we won't be sending them a packaged binary stream, we will send them an api request that contains various fields. this interface makes a huge (and wrong) assumption about how an email provider would be accepting data from us.

File: src/main/kotlin/io/zerobase/smarttracing/resources/Email.kt (hunk 20 above)
Comment (toadzky) [reply src/main/kotlin/io/zerobase/smarttracing/resources/Email.kt]: the name of the class say this is an email, but that's not what it is. this is a utility class that sends emails. an email has to/from addresses, a subject, a body, and optionally an attachment. this class has configuration on how to send an email.

again, as with my comments from above, i think we should separate out the various parts of sending an email:
- rendering a body
- rendering an attachment
- sending the email

we want an interface for the last one, as discussed elsewhere in this PR, that takes the parts of an email and lets the implementation decide how to package them together. we want the business logic to decide if an attachment is necessary and call the relevant attachment generation code with the correct arguments (this is what emily is working on). 

for the rendering of the body, since each message could
a) be template and each template has the possibility of different arguments and
b) the same message could be rendered differently for different notification mediums (text, email, etc) 
i think it makes sense for each message type to implement its own rendering. a good option, IMO, would be use kotlin sealed classes for this. here's an example of what it could look like:

```kotlin
sealed class UserNotification {
    abstract fun render(medium: NotificationMedium = NotificationMedium.EMAIL): String
}

class OrganizationOnboardingNotification(val organizationName: String): UserNotification {
    private val emailTemplate = "notifications/organization-onboarding/email.mustache"
    private val smsTemplate = "notifications/organization-onboarding/sms.mustache"

    override fun render(medium: NotificationMedium): String {
        // do the rendering
    }
}
```

File: src/main/kotlin/io/zerobase/smarttracing/resources/EmailSender.kt (hunk 21 above)
Comment (mbuchel) [reply src/main/kotlin/io/zerobase/smarttracing/resources/EmailSender.kt]: ok than in this case should emailsender not be on the email, and have amazonSES code inside the email sender?

**Review by toadzky — APPROVED** [review]
